src/storage/petastorm_storage_engine.py

```diff
@@ -0,0 +1,137 @@
+        """
+        Maintain a long live spark session and context.
+        """
+        self._spark = Session()
+        self.spark_session = self._spark.get_session()
+        self.spark_context = self._spark.get_context()
+
+    def _spark_url(self, table: DataFrameMetadata) -> str:
+        """
+        Generate a spark/petastorm url given a table
+        """
+        eva_dir = ConfigurationManager().get_value("core", "location")
+        output_url = os.path.join(eva_dir, table.name)
+
+        return Path(output_url).resolve().as_uri()
+
+    def create(self, table: DataFrameMetadata):
+        """
+        Create an empty dataframe in petastorm.
+        """
+        empty_rdd = self.spark_context.emptyRDD()
+
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            self.spark_session.createDataFrame(empty_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('overwrite') \
+                .parquet(self._spark_url(table))
+
+    def write_row(self, table: DataFrameMetadata, rows: []):
+        """
+        Write rows into the dataframe.
+
+        Arguments:
+            rows List[Dict{key, value}]: We can formally define a dataType Row.
+            keys within the Dict should be consistent with the table.schema.
+        """
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            rows_rdd = self.spark_context.parallelize(rows) \
+                .map(lambda x: dict_to_spark_row(table.schema.petastorm_schema,
+                                                 x))
+
+            self.spark_session.createDataFrame(rows_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('append') \
+                .parquet(self._spark_url(table))
+
+    def read(self, table: DataFrameMetadata) -> Iterator[Dict]:
+        """
+        Read the dataframe.
+
+        Return:
+            Iterator of Dict. Each Dict represents a Row.
+        """
+        with make_reader(self._spark_url(table)) as reader:
+            for row in reader:
+                yield row._asdict()
+
+    def read_lambda(self, table: DataFrameMetadata, fields: [
+                    str], predicate_func) -> Iterator[Dict]:
+        """
+        Read the rows that the predicate_func returns true on the given fields.
+
+        Argument:
+            fields List[str]: A list of fields (column names) to be considered.
+            predicate_func: customized function return bool
+        """
+        predicate = in_lambda(fields, predicate_func)
+        with make_reader(self._spark_url(table),
```

Comment:
same as the read

Reply:
Why have two methods in this case? why not just have `read` which takes optional params. `fields` and `predicate_func`. Technically if `fields` are not provided, shouldn't we take all the fields in the table?

Reply:
Makes sense.
Btw fields is something required by predicate_func